src/main/scala/com/twitter/finagle/HttpWebSocket.scala

```diff
@@ -47,103 +52,106 @@ trait WebSocketRichClient { self: Client[WebSocket, WebSocket] =>
+    }, addr, params)
   }
 
-  protected def copy1(
-    stack: Stack[ServiceFactory[WebSocket, WebSocket]] = this.stack,
-    params: Stack.Params = this.params
-  ): WebSocketClient = copy(stack, params)
+  protected def copy1(stack: Stack[ServiceFactory[WebSocket, WebSocket]] = this.stack,
+                      params: Stack.Params = this.params): WebSocketClient = copy(stack, params)
 
-  protected def newDispatcher(
-    transport: Transport[WebSocket, WebSocket] { type Context <: WebSocketClient.this.Context }
-  ): Service[WebSocket, WebSocket] = {
-    new SerialClientDispatcher(transport)
-  }
+  protected def newDispatcher(transport: Transport[WebSocket, WebSocket]
+    {type Context <: WebSocketClient.this.Context}): Service[WebSocket, WebSocket] = new SerialClientDispatcher(transport)
 
   def withTlsWithoutValidation(): WebSocketClient =
-    configured(Transport.ClientSsl(
-      Some(SslClientConfiguration(trustCredentials = TrustCredentials.Insecure))))
+    configured(Transport.ClientSsl(Some(SslClientConfiguration(trustCredentials = TrustCredentials.Insecure))))
+}
+
+object SessionIdleTimeout {
+  implicit val param: Stack.Param[SessionIdleTimeout] =
+    Stack.Param(SessionIdleTimeout(0))
+}
+
+case class SessionIdleTimeout(seconds: Int) {
+  def mk(): (SessionIdleTimeout, Stack.Param[SessionIdleTimeout]) =
+    (this, SessionIdleTimeout.param)
 }
 
 object WebSocketServer {
-  val stack: Stack[ServiceFactory[WebSocket, WebSocket]] =
-    StackServer.newStack
+  val stack: Stack[ServiceFactory[WebSocket, WebSocket]] = StackServer.newStack
 }
 
-case class WebSocketServer(
-  stack: Stack[ServiceFactory[WebSocket, WebSocket]] = WebSocketServer.stack,
-  params: Stack.Params = StackServer.defaultParams + ProtocolLibrary("websocket"),
-  sessionIdleTimeout: Int = 0
-) extends StdStackServer[WebSocket, WebSocket, WebSocketServer] {
+case class WebSocketServer(stack: Stack[ServiceFactory[WebSocket, WebSocket]] = WebSocketServer.stack,
+                           params: Stack.Params = StackServer.defaultParams + ProtocolLibrary("websocket"),
+                           sessionIdleTimeout: Int = 0) extends StdStackServer[WebSocket, WebSocket, WebSocketServer] {
   protected type In = WebSocket
   protected type Out = WebSocket
   protected type Context = TransportContext
 
   protected def newListener(): Listener[WebSocket, WebSocket, TransportContext] = {
-    val Label(label) = params[Label]
-    val pipeline = WebSocketCodec(sessionIdleTimeout)
-      .server(ServerCodecConfig(label, new SocketAddress {}))
-      .pipelineFactory
-
-    Netty3Listener(pipeline, params)
+    val Label(_) = params[Label]
+    Netty4Listener((pipeline: ChannelPipeline) => {
+      pipeline.addLast("httpCodec", new HttpServerCodec)
+      pipeline.addLast("httpAggregator", new HttpObjectAggregator(65536))
+      pipeline.addLast("idleStateHandler",  new IdleStateHandler(0, 0, params[SessionIdleTimeout].seconds))
```

Comment:
If the `SessionIdleTimeout` is not set we would not need to add this handler. Is it possible to check that before adding?

Reply:
The default value for `SessionIdleTimeout` 0 mean the same behaviour.
But as you suggest we could completely ignore it when `SessionIdleTimeout` is 0 .